app/src/main/java/com/github/polybooks/core/database/implementation/SaleDatabase.kt

```diff
@@ -0,0 +1,173 @@
+package com.github.polybooks.core.database.implementation
+
+import com.github.polybooks.core.*
+import com.github.polybooks.core.database.DatabaseException
+import com.github.polybooks.core.database.interfaces.SaleDatabase
+import com.github.polybooks.core.database.interfaces.SaleFields
+import com.github.polybooks.core.database.interfaces.SaleOrdering
+import com.github.polybooks.core.database.interfaces.SaleQuery
+import com.google.firebase.firestore.CollectionReference
+import com.google.firebase.firestore.FirebaseFirestore
+import com.google.firebase.firestore.Query
+import com.google.firebase.firestore.QueryDocumentSnapshot
+import java.util.*
+import java.util.concurrent.CompletableFuture
+
+class SaleDatabase : SaleDatabase {
+
+    private val db: FirebaseFirestore = FirebaseFirestore.getInstance()
+    private val saleRef: CollectionReference = db.collection(getCollectionName())
+
+    inner class SalesQuery : SaleQuery {
+
+        private var isbn13: String? = null
+        private var title: String? = null
+
+        private var interests: Set<Interest>? = null
+        private var states: Set<SaleState>? = null
+        private var conditions: Set<BookCondition>? = null
+
+        private var minPrice: Float? = null
+        private var maxPrice: Float? = null
+
+        override fun onlyIncludeInterests(interests: Collection<Interest>): SaleQuery {
+            if (!interests.isEmpty()) this.interests = interests.toSet()
+            return this
+        }
+
+        override fun searchByTitle(title: String): SaleQuery {
```

Comment:
Should you not set ```interests``` and ```isbn``` to null ?

Reply:
I mean, I guess it depends. This is the discussion we had a few days ago. Namely, do we complexify the code to make the search "logical", or let the user do whatever he wants. Both could work imo